File: android/app/src/main/java/com/murmur/app/domain/aggregation/SessionAggregator.kt
package com.murmur.app.domain.aggregation

import com.murmur.app.domain.model.DetectedSession
import com.murmur.app.domain.model.LedgerEntry
import com.murmur.app.domain.model.SessionStatus
import com.murmur.app.domain.model.SourceKind

/**
 * Cross-platform session aggregation with shared algorithms.
 * mergeAdjacent, interval union, gross vs deduped active_seconds,
 * tool/source distribution, completion rate.
 */
object SessionAggregator {

    fun isIncluded(session: DetectedSession): Boolean {
        return when (session.status) {
            SessionStatus.PENDING, SessionStatus.COMPLETED, SessionStatus.SUSPECTED -> true
            else -> false
        }
    }

    fun normalizeSessions(sessions: List<DetectedSession>): List<DetectedSession> {
        return sessions.filter { isIncluded(it) }
    }

    // MARK: - Adjacent Merge

    /**
     * Merge sessions of same tool within window (default 180s).
     * Same device, same source_platform, same source_kind, same tool_id.
     */
    fun mergeAdjacentSessions(
        sessions: List<DetectedSession>,
        windowSeconds: Long = 180
    ): List<DetectedSession> {
        val included = normalizeSessions(sessions)
        if (included.size <= 1) return included.toList()

        val sorted = included.sortedBy { it.startedAt }
        val merged = mutableListOf<DetectedSession>()
        var current = sorted[0]

        for (i in 1 until sorted.size) {
            val next = sorted[i]
            val sameTool = current.toolId == next.toolId
            val sameSource = current.sourcePlatform == next.sourcePlatform &&
                    current.sourceKind == next.sourceKind
            val gap = (next.startedAt - current.endedAt) / 1000

            if (sameTool && sameSource && gap in 1..windowSeconds) {
                current = current.copy(
                    endedAt = maxOf(current.endedAt, next.endedAt),
                    activeSeconds = current.activeSeconds + next.activeSeconds,
                    promptCount = (current.promptCount ?: 0) + (next.promptCount ?: 0),
                    updatedAt = System.currentTimeMillis()
                )
            } else {
                merged.add(current)
                current = next
            }
        }
        merged.add(current)
        return merged
    }

    // MARK: - Active Seconds

    fun calculateGrossActiveSeconds(sessions: List<DetectedSession>): Long {
        return normalizeSessions(sessions).sumOf { it.activeSeconds }
    }

    /**
     * Deduped active seconds: interval union of all session time ranges.
     */
    fun calculateDedupedActiveSeconds(sessions: List<DetectedSession>): Long {
        val included = normalizeSessions(sessions).filter { it.activeSeconds >= 15 }
        if (included.isEmpty()) return 0

        val intervals = included.map {
            it.startedAt.toDouble() to it.endedAt.toDouble()
        }.sortedBy { it.first }

        val merged = mutableListOf<Pair<Double, Double>>()
        var current = intervals[0]

        for (i in 1 until intervals.size) {
            if (intervals[i].first > current.second) {
                merged.add(current)
                current = intervals[i]
            } else {
                current = current.copy(second = maxOf(current.second, intervals[i].second))
            }
        }
        merged.add(current)

        return merged.sumOf { (it.second - it.first).toLong() / 1000 }
    }

    // MARK: - Distribution

    fun calculateToolDistribution(sessions: List<DetectedSession>): Map<String, Long> {
        val included = normalizeSessions(sessions)
        val dist = mutableMapOf<String, Long>()
        for (s in included) {
            val name = s.toolName.ifEmpty { s.toolId }
            dist[name] = (dist[name] ?: 0) + s.activeSeconds
        }
        return dist.toList().sortedByDescending { it.second }.toMap()
    }

    fun calculateAppWebBreakdown(sessions: List<DetectedSession>): Pair<Long, Long> {
        val included = normalizeSessions(sessions)
        var app = 0L
        var web = 0L
        for (s in included) {
            if (s.sourceKind == SourceKind.WEB) web += s.activeSeconds
            else app += s.activeSeconds
        }
        return app to web
    }

    // MARK: - Completion Rate

    fun calculateCompletionRate(sessions: List<DetectedSession>, entries: List<LedgerEntry>): Double {
        val included = normalizeSessions(sessions)
        val pending = included.count { it.status == SessionStatus.PENDING || it.status == SessionStatus.SUSPECTED }
        val completed = included.count { it.status == SessionStatus.COMPLETED }
        val total = pending + completed
        return if (total > 0) completed.toDouble() / total else 0.0
    }
}

File: macos/Murmur/Aggregation/SessionAggregator.swift
import Foundation

/// Cross-platform session aggregation logic.
/// Shared algorithm: merge adjacent, interval union, gross vs deduped active_seconds,
/// tool/source distribution, completion rate.
struct SessionAggregator {

    // MARK: - Filtering

    static func isIncluded(_ session: DetectedSession) -> Bool {
        switch session.status {
        case .pending, .completed, .suspected: return true
        case .ignored, .merged: return false
        }
    }

    static func normalizeSessions(_ sessions: [DetectedSession]) -> [DetectedSession] {
        return sessions.filter(isIncluded)
    }

    // MARK: - Adjacent Merge

    /// Merge sessions of same tool within window (default 180s).
    /// Same device, same source_platform, same source_kind, same tool_id.
    static func mergeAdjacentSessions(_ sessions: [DetectedSession], windowSeconds: TimeInterval = 180) -> [DetectedSession] {
        let included = normalizeSessions(sessions)
        guard included.count > 1 else { return included }

        let sorted = included.sorted { $0.startedAt < $1.startedAt }
        var merged: [DetectedSession] = []
        var current = sorted[0]

        for i in 1..<sorted.count {
            let next = sorted[i]
            let sameTool = current.toolId == next.toolId
            let sameSource = current.sourcePlatform == next.sourcePlatform &&
                             current.sourceKind == next.sourceKind
            let gap = next.startedAt.timeIntervalSince(current.endedAt)

            if sameTool && sameSource && gap > 0 && gap <= windowSeconds {
                var updated = current
                updated.endedAt = max(current.endedAt, next.endedAt)
                updated.activeSeconds = current.activeSeconds + next.activeSeconds
                updated.promptCount = (current.promptCount ?? 0) + (next.promptCount ?? 0)
                updated.updatedAt = Date()
                current = updated
            } else {
                merged.append(current)
                current = next
            }
        }
        merged.append(current)
        return merged
    }

    // MARK: - Active Seconds

    /// Gross active seconds: simple sum of all included sessions.
    static func calculateGrossActiveSeconds(_ sessions: [DetectedSession]) -> Int {
        return normalizeSessions(sessions).reduce(0) { $0 + $1.activeSeconds }
    }

    /// Deduped active seconds: interval union of all session time ranges.
    static func calculateDedupedActiveSeconds(_ sessions: [DetectedSession]) -> Int {
        let included = normalizeSessions(sessions).filter { $0.activeSeconds >= 15 }
        guard !included.isEmpty else { return 0 }

        let intervals = included.map {
            (start: $0.startedAt.timeIntervalSince1970, end: $0.endedAt.timeIntervalSince1970)
        }.sorted { $0.start < $1.start }

        var merged: [(start: TimeInterval, end: TimeInterval)] = []
        var current = intervals[0]

        for i in 1..<intervals.count {
            if intervals[i].start > current.end {
                merged.append(current)
                current = intervals[i]
            } else {
                current.end = max(current.end, intervals[i].end)
            }
        }
        merged.append(current)

        return Int(merged.reduce(0) { $0 + ($1.end - $1.start) })
    }

    // MARK: - Distribution

    static func calculateToolDistribution(_ sessions: [DetectedSession]) -> [(toolName: String, seconds: Int)] {
        let included = normalizeSessions(sessions)
        var dist: [String: Int] = [:]
        for s in included {
            let name = s.toolName ?? s.toolId ?? "Unknown"
            dist[name, default: 0] += s.activeSeconds
        }
        return dist.map { ($0.key, $0.value) }.sorted { $0.1 > $1.1 }
    }

    static func calculateAppWebBreakdown(_ sessions: [DetectedSession]) -> (app: Int, web: Int) {
        let included = normalizeSessions(sessions)
        var app = 0, web = 0
        for s in included {
            if s.sourceKind == .web { web += s.activeSeconds }
            else { app += s.activeSeconds }
        }
        return (app, web)
    }

    // MARK: - Completion Rate

    static func calculateCompletionRate(sessions: [DetectedSession], entries: [LedgerEntry]) -> Double {
        let included = normalizeSessions(sessions)
        let pendingCount = included.filter { $0.status == .pending || $0.status == .suspected }.count
        let completedCount = included.filter { $0.status == .completed }.count
        let total = pendingCount + completedCount
        return total > 0 ? Double(completedCount) / Double(total) : 0
    }

    // MARK: - Daily Summary

    static func buildDailySummary(sessions: [DetectedSession], entries: [LedgerEntry], localDate: String) -> DailySummary {
        let dateSessions = sessions.filter { $0.localDate == localDate }
        let dateEntries = entries.filter { $0.localDate == localDate }
        let included = normalizeSessions(dateSessions)
        let (appSec, webSec) = calculateAppWebBreakdown(included)
        let promptTotal = included.reduce(0) { $0 + ($1.promptCount ?? 0) }

        let gross = calculateGrossActiveSeconds(included)
        return DailySummary(
            id: localDate,
            localDate: localDate,
            detectedSessionCount: included.count,
            pendingSessionCount: included.filter { $0.status == .pending }.count,
            completedSessionCount: included.filter { $0.status == .completed }.count,
            detectedActiveSeconds: gross,
            grossActiveSeconds: gross,
            dedupedActiveSeconds: calculateDedupedActiveSeconds(included),
            appActiveSeconds: appSec,
            webActiveSeconds: webSec,
            promptCount: promptTotal,
            completionRate: calculateCompletionRate(sessions: dateSessions, entries: dateEntries)
        )
    }
}
